### PR TITLE
tedge mapper dynamic discovery of operations

### PR DESCRIPTION
**Pull request by initard**
## Proposed changes

Implementing fs_notify crate for c8y configuration plugin.
This commit contains the changes of #1244 applied on top of it. When reviewing please review only the last commit: d737517d97837d3fffd64ba6d5b900297e4748e0

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
- #1246

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by didier-wenzek — COMMENTED** [review]
This PR needs to be updated now that `fs-notify` has been merged

File: crates/common/mqtt_channel/src/errors.rs [hunk 1]
```diff
@@ -38,6 +38,9 @@ pub enum MqttError {
         "The send channel of the connection has been closed and no more messages can be published"
     )]
     SendOnClosedConnection,
+
+    #[error(transparent)]
+    FromInotifyError(#[from] tedge_utils::fs_notify::NotifyStreamError),
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/common/mqtt_channel/src/errors.rs:43]: To be removed

File: crates/core/tedge_mapper/src/c8y/dynamic_discovery.rs [hunk 2]
```diff
@@ -19,81 +21,54 @@ pub struct DiscoverOp {
 #[derive(thiserror::Error, Debug)]
 #[allow(clippy::enum_variant_names)]
 pub enum DynamicDiscoverOpsError {
-    #[error("Failed to add watch to directory: {0}")]
-    FailedtoAddWatch(String),
-
-    #[error("A non-UTF8 name cannot be used as an operation name: {0:?}")]
-    NotAnOperationName(OsString),
+    #[error("A non-UTF8 path cannot be parsed as an operation: {0:?}")]
+    NotAnOperation(PathBuf),
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/dynamic_discovery.rs:25]: Why changing the error message?

I think that the former message makes more sense for the end user. Parsing failure is a concern for the developer not for the user.

File: crates/core/tedge_mapper/src/c8y/dynamic_discovery.rs [hunk 3]
```diff
@@ -19,81 +21,54 @@ pub struct DiscoverOp {
...
-pub fn create_inotify_watch(ops_dir: PathBuf) -> Result<Inotify, DynamicDiscoverOpsError> {
-    let mut inotify = Inotify::init()?;
-    inotify
-        .add_watch(ops_dir.clone(), WatchMask::CLOSE_WRITE | WatchMask::DELETE)
-        .map_err(|_| {
-            DynamicDiscoverOpsError::FailedtoAddWatch(ops_dir.to_string_lossy().to_string())
-        })?;
-    Ok(inotify)
-}
+/// depending on which editor you use, temporary files could be created.
```
[the remark this answers is not in the document]
Comment (didier-wenzek) [reply crates/core/tedge_mapper/src/c8y/dynamic_discovery.rs:31]: Because some editors, if not all, create temporary files in the watched directory and we don't want  to trigger events for those.

@initard Maybe the comment might be improved since not understood by all.

File: crates/core/tedge_mapper/src/core/mapper.rs [hunk 4]
```diff
@@ -188,21 +178,15 @@ async fn process_inotify_and_mqtt_messages(
                     }
                 }
             }
-        }, // On error continue to process only mqtt messages.
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/core/mapper.rs:191]: I would keep this comment.

**Review by didier-wenzek — COMMENTED** [review]
There is a clippy warning to be fixed.

**Review by rina23q — COMMENTED** [review]
The function `operation_name_is_valid` works totally fine, however, there is a bug in upper level.

1. Subscribe `c8y/s/us`, e.g. 
```
tedge mqtt sub "c8y/s/us
```
2. Create a file with tilda, e.g. run 
```
sudo touch /etc/tedge/operations/c8y/c8y_Test~
```
3. In this case, you can't see anything on the 1) subscriber.
4. Create a file with ASCII, e.g. run
```
sudo touch /etc/tedge/operations/c8y/c8y_Go
```
5. Check the subscriber 1). You see `c8y_Test~` is declared as supported operation.
```
[c8y/s/us] 114,c8y_DownloadConfigFile,c8y_SoftwareUpdate,c8y_LogfileRequest,c8y_Go,c8y_Restart,c8y_UploadConfigFile,c8y_Test~
```

File: crates/core/tedge_mapper/src/core/mapper.rs [hunk 5]
```diff
@@ -86,7 +87,7 @@ impl Mapper {
         }
     }
 
-    pub(crate) async fn run(&mut self, ops_dir: Option<PathBuf>) -> Result<(), MqttError> {
+    pub(crate) async fn run(&mut self, ops_dir: Option<&Path>) -> Result<(), MqttError> {
```
Comment (matthiasbeyer) [inline crates/core/tedge_mapper/src/core/mapper.rs:90]: This change seems to be unrelated to the scope of this PR, or am I missing something?

File: crates/core/tedge_mapper/src/core/mapper.rs [hunk 6]
```diff
@@ -102,7 +103,7 @@ impl Mapper {
     }
 
     #[instrument(skip(self), name = "messages")]
-    async fn process_messages(&mut self, ops_dir: Option<PathBuf>) -> Result<(), MqttError> {
+    async fn process_messages(&mut self, ops_dir: Option<&Path>) -> Result<(), MqttError> {
```
Comment (matthiasbeyer) [inline crates/core/tedge_mapper/src/core/mapper.rs:106]: Same as above.

**Review by matthiasbeyer — COMMENTED** [review]
I have some comments, hope you don't mind! :smile:

File: crates/core/c8y_smartrest/src/operations.rs [hunk 7]
```diff
@@ -48,6 +48,16 @@ pub struct Operations {
...
 }
 
+/// depending on which editor you use, temporary files could be created that contain the name of
+/// the file.
+/// this `operation_name_is_valid` fn will ensure that only files that do not contain
+/// any special characters are allowed.
+pub fn is_valid_operation_name(operation: &str) -> bool {
+    operation
+        .chars()
+        .all(|c| c.is_ascii_alphabetic() | c.is_numeric() | c.eq(&'_'))
```
Comment (rina23q) [inline crates/core/c8y_smartrest/src/operations.rs:58]: A bit surprised that this function rejects even `-`.

**Review by rina23q — APPROVED** [review]
[The bug](https://github.com/thin-edge/thin-edge.io/pull/1250#pullrequestreview-1043737985) was fixed!

File: crates/core/c8y_smartrest/src/operations.rs (hunk 7 above)
Comment (initard) [reply crates/core/c8y_smartrest/src/operations.rs:58]: yes `_` is not alphabetic() or numeric() and our operations are prefixed with `c8y_`

Comment (initard) [reply crates/core/c8y_smartrest/src/operations.rs:58]: ohh thanks!
